docs/decisions/0045-function-call-behavior.md

```diff
@@ -0,0 +1,326 @@
+- It should be possible to override the prompt execution settings specified in the prompt by using the prompt execution settings specified in the code.
+
+## Existing function calling behavior model - ToolCallBehavior
+Today, SK uses the `ToolCallBehavior` abstract class and its derivatives `KernelFunctions`, `EnabledFunctions`, and `RequiredFunction` to define function calling behavior for the OpenAI connector. The behavior is specified via the `OpenAIPromptExecutionSettings.ToolCallBehavior` property. The model is identical for other connectors and differs only in the function call behavior class names.
+
+```csharp
+OpenAIPromptExecutionSettings settings = new() { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
+
+or
+
+GeminiPromptExecutionSettings settings = new() { ToolCallBehavior = GeminiToolCallBehavior.AutoInvokeKernelFunctions };
+```
+
+Taking into account that the function-calling behavior has been around since SK v1 release and might be used extensively as a result, the new function-calling abstraction must be introduced and coexist alongside the existing function-calling model. This will prevent breaking changes and allow consumers to gradually migrate from the current model to the new one.
+
+## [New model] Option 1.1 - A class per function call choice
+To satisfy the "no breaking changes" requirement above and the "connector/mode-agnostic model" decision driver, the new set of connector agnostic classes needs to be introduced.
+
+### Function call choice classes 
+The `FunctionChoiceBehavior` class is abstract base class for all *FunctionChoiceBehavior classes:
+
+```csharp
+public abstract class FunctionChoiceBehavior
+{
+    public static FunctionChoiceBehavior AutoFunctionChoice(IEnumerable<KernelFunction>? functions = null, bool autoInvoke = true) { ... }
+    public static FunctionChoiceBehavior RequiredFunctionChoice(IEnumerable<KernelFunction>? functions = null, bool autoInvoke = true) { ... }
+    public static FunctionChoiceBehavior None { get { ... }; }
+
+    public abstract FunctionChoiceBehaviorConfiguration GetConfiguration(FunctionChoiceBehaviorContext context);
+}
+```
+
+All the `FunctionChoiceBehavior` derivatives must implement the abstract `GetConfiguration` method, which is called with the `FunctionChoiceBehaviorContext` provided by connectors. This method returns the `FunctionChoiceBehaviorConfiguration` back to the connectors, instructing them to behave in a specific way described by a particular function call choice behavior class with respect to function calling.
+
+```csharp
+public class FunctionChoiceBehaviorContext
+{
+    public Kernel? Kernel { get; init; }
+}
+
+public class FunctionCallChoiceConfiguration
+{
+    public IEnumerable<KernelFunctionMetadata>? AvailableFunctions { get; init; }
+    public IEnumerable<KernelFunctionMetadata>? RequiredFunctions { get; init; }
+    public bool? AllowAnyRequestedKernelFunction { get; init; }
```

Comment:
Should this not be something like this

```csharp
public IEnumerable<KernelFunctionMetadata>? Functions { get; init; } // Functions to advertise to the LLM
public bool? IsRequired { get; init; } // True if the LLM must call one of the functions
public bool? IsNone { get; init; } // True if the LLM must not call one of the functions but instead should respond with a message
```

Reply:
The ADR is slightly behind the PR - https://github.com/microsoft/semantic-kernel/pull/6083, in which the class has been changed based on the latest feedback to:
<img width="767" alt="image" src="https://github.com/microsoft/semantic-kernel/assets/68852919/e082da94-3212-4c8c-8499-6b4d67e8412e">

I will update the ADR later to reflect the latest state of that PR.

Reply:
Shouldn't the ADR be driving what is done in code, not the other way around? I think it makes it more difficult for other SK languages to fully understand the scope of the work if the code is first completed, and then the ADR is updated.

Reply:
It should, but because of the complexity and many comments, despite the ADR presented and reviewed, the code evolved much faster and became a driver for the ADR.